### ISSUE TITLE
SSE: possible standards compliance error with "missing ACK" (+ solution)

### ISSUE THREAD

**Issue by philbowles** [me-no-dev/ESPAsyncWebServer#765]
I have code which sends multiple SSE events to a webpage. It is clear that both Chrome and Firefox do not always send a TCP ACK for every instance of an event sent by the library.
![image](https://user-images.githubusercontent.com/27769598/82340558-a7393880-99ef-11ea-9fc2-427ab1629286.png)

Unfortunately, the default timeout action when ACK not received is to simply close the connection, which "breaks" the web interface as the browser  javascript correctly "errors" the connection, when it *could* ("should" ?) quite easily continue, since the message has actually been successfully sent and acted upon by the browser:

![image](https://user-images.githubusercontent.com/27769598/82340867-039c5800-99f0-11ea-8a0d-6f7097902128.png)

This is also a good example of why I think the examples in your documentation (and any production code) should *not* use millis() as the ID fields, as it is clear from the above that ID collisions are not only possible, but inevitable. I would recommend a unique incrementing reference ID per message.
 
I'm not a standards expert, but closing the connection on "missing ACK" this *seems* to contradict RFC 2616

> 8.2.1 Persistent Connections and Flow Control
> HTTP/1.1 servers SHOULD maintain persistent connections and use TCP's flow control mechanisms to resolve temporary overloads, rather than terminating connections with the expectation that clients will retry. The latter technique can exacerbate network congestion

Either way, it causes major problems for the user and there is an "easy fix" (2 lines) which removes the problem completely.

.I accept that the full RFC recommendation of "exponential fallback" may be over-the top, but there is a much simpler solution. Even if it is both Chrome and Firefox being "in error" my solution fails gracefully and allows continuation with no side effects. If a "missing ACK" is indeed something SSE servers should support then my fix renders the library compliant.

I have made small changes which demonstrate the problem with an easy MCVE.
1)  addition of "dump" function in AsyncEventSourceMessage (demo only, not suggested for use in production)

```cpp
    void dump(){ Serial.printf("T=%u: ACK TIMEOUT *********** %s",millis(),_data); } // debug only
```

2) very small change to AsyncEventSourceClient

```cpp
void AsyncEventSourceClient::_onTimeout(uint32_t time __attribute__((unused))){
//  _client->close(true); - Too harsh / stds compilant??? Either way, its  breaks webui
    _messageQueue.front()->dump(); // debug only
    _messageQueue.remove(_messageQueue.front()); // much softer
    _runQueue(); // everything carries on working
}
```
My reasoning is as follows:

1. - the message has actually been sent - all thats missing is the ACK
2. - _onTimeout can only ever be called if there is a unacked msg at head of queue so no error-checking is required
3. - without this, the queue just keeps growing till it breaks the limit, which - in my view - is not a good or easily-manageable outcome
4. - it does no harm to any other code
5. - it does not require any changes to existing user code either in the ESP or the web UI js

# The MCVE 

```cpp
#include<ESP8266WiFi.h>
#include<ESPAsyncWebServer.h>
#include<Ticker.h>

AsyncWebServer server(80);
AsyncEventSource events("/events");

Ticker  T;

char home[]="<!DOCTYPE html><html><head><script>"\
"document.addEventListener(\"DOMContentLoaded\", function() {"\
"    source=new EventSource(\"/events\");"\
"    source.onmessage=function(e){ console.warn(\"message \",e.data); };"\
"    source.onerror=function(e){ console.warn(\"SSE ERROR!\",e); };"\        
"    } "\
");"\
"</script></head><body>SSE</body></html>";

void setup(){
  Serial.begin(115200);
  WiFi.begin("XXXXXXXX", "XXXXXXXX");
  WiFi.waitForConnectResult();
  
  Serial.println("Ready");
  Serial.print("IP address: ");
  Serial.println(WiFi.localIP());

  events.onConnect([](AsyncEventSourceClient *client){
    if(client->lastId()){
      Serial.printf("Client reconnected! Last message ID that it got is: %u\n", client->lastId());
    }
    client->send("hello!",NULL,1,5000);
    events.send("rapid-fire 1",NULL,2);
    events.send("rapid-fire 2",NULL,3);
  });
  server.addHandler(&events);
  server.on("/",[](AsyncWebServerRequest* p){ p->send(200,"text/html",home); });
  server.begin();
  T.attach_ms(1000,[]{ events.send("ping",NULL,millis()); });
}

void loop(){
}
```

# Sample Output

```
Ready
IP address: 192.168.1.104
T=18335: ACK TIMEOUT *********** retry: 5000
id: 1
data: hello!
```

BUT - the web UI continues glitch-free:
![image](https://user-images.githubusercontent.com/27769598/82344507-6a237500-99f4-11ea-8984-57fb7b58f55b.png)

**Comment by iafilius** [comment]
Hi Phil,

Nice discovery, but i think you're not completely right in reasoning.

On tcp level, an ack isn't needed to be send for every packet. 
simplified: sending an ack for the latest packet within the windows ack's all previous packets. 
PS: I i know there is also Selective ACK (SACK) en such complicate the discussion a lot, but i guess that's not used. can't check your capture as you shared the screenshot of wireshark, not the data.

So there is no (guaranteed) 1 to 1 relation to #sending packets and #receiving packets.
So if the ack is received, it acknowledges the previous data withing the windows, and that might include another event.
That data should be found and set to 'processed', and removed from queue.

I'm not entirely sure myself is the asyncevensource onAck is one-to-one coupled to the tcpstack ACK, you checked this i guess? (i not now/yet)
```
void AsyncEventSourceClient::_onAck(size_t len, uint32_t time){
  while(len && !_messageQueue.isEmpty()){
    len = _messageQueue.front()->ack(len, time);
    if(_messageQueue.front()->finished())
      _messageQueue.remove(_messageQueue.front());
  }

  _runQueue();
}
```


as a solution waiting on a timeout and then remove it from the queue kinda works, certain better then just resetting tcp session.

so the timeout on asynceventsource in :
```
void AsyncEventSourceClient::_onTimeout(uint32_t time __attribute__((unused))){
  _client->close(true);
}
```
should not happen when the event was nicely acknowledged on tcp level.

So i think if all above is true, the real solution should be to fix the relation between tcp ack's and asynceventsource event id's-ack's 


## Regarding the millis as unique id, i think you're right. but the code uses even the id a defined as 'uint32_t time' t in:
```
void AsyncEventSourceClient::_onAck(size_t len, uint32_t time){
  while(len && !_messageQueue.isEmpty()){
    len = _messageQueue.front()->ack(len, time);
    if(_messageQueue.front()->finished())
      _messageQueue.remove(_messageQueue.front());
  }

  _runQueue();
}
```
to find the particular message back. so it's not just the examples.
if it is a unique_id, also call it that way, and not time, although i see why millis(time) fits perfectly well in low-rate conditions.

besides... I didn't see any activity from @me-no-dev for about 8 months on this project, quite a bit merge requests waiting. is it abandoned?/forked?

@philbowles  as you're using SSE, fast rates, and maybe even bad wifi connection, you probably want to have a look at my merge request regarding disabeling disabeling re-entrant runqueue for stability reasons (maybe not nicest implementation, but improves stability a lot for me)


Regards Arjan
